src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file thermal_functional_material.hpp
+ *
+ * @brief The material and load types for the thermal functional physics module
+ */
+
+#pragma once
+
+#include "serac/numerics/functional/functional.hpp"
+
+/// ThermalConductionFunctional helper structs
+namespace serac::Thermal {
+
+/// Linear isotropic thermal conduction material model
```

Comment:
It looks like the relevant properties are assumed to be constant over the domain (or at least over the material). Is it possible to instead define the properties via Coefficient?

Reply:
Follow up question, are tensor-valued conductivities on the menu?

Reply:
Unfortunately no. The coefficient virtual function interface is not amenable to GPUs. We can add a `FiniteElementState`-type interface or a CPU-only coefficient style one. Any thoughts @samuelpmishLLNL ?

Reply:
We can easily implement a tensor valued conductivity using this interface.

Reply:
We would probably be pushing `FiniteElementState`s into `GridFunctionCoefficient`s anyway, so I think either of these approaches would work, but we'd definitely want a way to provide spatially-varying property fields (and loads, etc.) as a way to inject design, uncertainty, and so on.

Reply:
I don't see any reason why a general anisotropic linear model can't be implemented immediately as:

```cpp
template < int dim >
struct LinearConductor {
   double density_;
   double specific_heat_capacity_;
   tensor<double, dim, dim> thermal_conductivity_;

   template <typename T1, typename T2>
   SERAC_HOST_DEVICE auto operator()(T1& /* temperature */, T2& temperature_gradient) const
   {
     return -thermal_conductivity_ * temperature_gradient;
   }
};
```

(note: this implementation is a homogeneous material)

If you wanted a spatially-varying thermal conductivity tensor, you could introduce it by passing a `QuadratureData< tensor<double, dim, dim> >` into a call to `Functional::AddDomainIntegral`

Reply:
> we'd definitely want a way to provide spatially-varying property fields (and loads, etc.) as a way to inject design, uncertainty, and so on.

If that's your goal, it seems more natural (to me) to express that with the multi-trial space flavor of Functional, something like

```cpp
template < int dim >
struct ParametrizedLinearConductor {
   double density_;
   double specific_heat_capacity_;
   
   tensor< double, dim, dim > some_parametrized_function(double design_parameter) {
       // you implement this
   }

   template <typename T1, typename T2>
   SERAC_HOST_DEVICE auto operator()(T1& /* temperature */, T2& temperature_gradient, T3 & design_parameter) const
   {
     tensor< double, dim, dim > thermal_conductivity = some_parametrized_function(design_parameter);
     return -thermal_conductivity_ * temperature_gradient;
   }
};
```

Reply:
This could work in at least some cases, but I think it violates our separation of concerns, in that I would need a serac object to know about design parameters, and would need to derive different serac objects depending on what I want to do in LiDO , e.g.,  am I interpolating properties based on design fields? Am I just assembling a spatially heterogeneous response due to different material regions? Am I passing you instances of a stochastic property field? Whereas all these use cases would be addressed if I can pass a FiniteElementState representation of the field.

Reply:
I didn't explain clearly in my previous comment, but the design parameter in this snippet:
```cpp
template < int dim >
struct ParametrizedLinearConductor {
   double density_;
   double specific_heat_capacity_;
   
   tensor< double, dim, dim > some_parametrized_function(double design_parameter) {
       // you implement this
   }

   template <typename T1, typename T2>
   SERAC_HOST_DEVICE auto operator()(T1& /* temperature */, T2& temperature_gradient, T3 & design_parameter) const
   {
     tensor< double, dim, dim > thermal_conductivity = some_parametrized_function(design_parameter);
     return -thermal_conductivity_ * temperature_gradient;
   }
};
```
is a finite element field evaluated at a particular quadrature point. If you want a nodally interpolated design parameter, then you use the H1 function space, if you want discontinuous, elementwise values, then L2 may be more appropriate.

The type `ParametrizedLinearConductor` is just expressing how that parameter controls the physical response of your material at a given quadrature point, not how that quadrature point value of the design parameter is calculated.

Reply:
After talking with @jamiebramwell , I believe this interface will work for LiDO's needs.